modules/nw-http2-haproxy.adoc

```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * networking/ingress-operator.adoc
+
+[id="nw-http2-haproxy{context}"]
+= HTTP/2 Ingress connectivity
+
+You can enable transparent end-to-end HTTP/2 connectivity in HAProxy. It allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.
+
+You can enable this HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.
+
+To enable the use of HTTP/2 for the connection from the client to HAProxy, a
+route must specify a custom certificate. A route that uses the default
+certificate cannot use HTTP/2. This restriction is necessary to avoid problems
+from connection coalescing, where the client re-uses a connection for different
+routes that use the same certificate.
+
+The connection from HAProxy to the application Pod can use HTTP/2 only for
+re-encrypt routes and not for edge-terminated or insecure routes. This
+restriction comes from the fact that HAProxy uses Application-Level Protocol
+Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2
+with the back-end. The implication is that end-to-end HTTP/2 is possible with
+passthrough and re-encrypt and not with insecure or edge-terminated routes.
+
+[IMPORTANT]
+====
+A connection that uses the HTTP/2 protocol cannot be upgraded to the WebSocket protocol. If you have a back-end application that is designed to allow WebSocket connections, it must not allow a connection to negotiate use of the HTTP/2 protocol or else WebSocket connections will fail.
```

Comment:
Iterating on this, we could use the following if we want to be more explicit and it's not too wordy:

> For non-passthrough routes, the IngressController negotiates its connection to the application independently of the connection from the client.  This means a client may connect to the IngressController and negotiate HTTP/1.1, and the IngressController may then connect to the application, negotiate HTTP/2, and forward the request from the client HTTP/1.1 connection using the HTTP/2 connection to the application.  This poses a problem if the client subsequently tries to upgrade its connection from HTTP/1.1 to the WebSocket protocol, because the IngressController cannot forward WebSocket to HTTP/2 and cannot upgrade its HTTP/2 connection to WebSocket.  Consequently, if you have an application that is intended to accept WebSocket connections, it must not allow negotiating the HTTP/2 protocol or else clients will fail to upgrade to the WebSocket protocol.